[profile] Implement a non-mmap path when reading profile files from a non-local filesystem

w2yehia opened the pull request:
On AIX, when accessing mmap'ed memory associated to a file on NFS, a SIGBUS might be raised at random.
The problem is still in open state with the OS team.

This PR teaches the profile runtime, under certain conditions, to avoid the mmap when reading the profile file during online merging.
This PR has no effect on any platform other than AIX because I'm not aware of this problem on other platforms.
Other platforms can easily opt-in to this functionality in the future.

The logic in function `is_local_filesystem` was copied from [llvm/lib/Support/Unix/Path.inc](https://github.com/llvm/llvm-project/blob/f388ca3d9d9a58e3d189458b590ba68dfd9e5a2d/llvm/lib/Support/Unix/Path.inc#L515) (https://reviews.llvm.org/D58801), because it seems that the compiler-rt/profile cannot reuse code from llvm except through `InstrProfData.inc`.

Thanks to @hubert-reinterpretcast for substantial feedback downstream.

Commit:
[profile] Implement a non-mmap path when reading profile files from a non-local filesystem

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,113 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+    PROF_ERR("%s: malloc failed: %s\n", __func__, strerror(errno));
+    return;
+  }
+  if (fseek(F, 0L, SEEK_SET) != 0) {
```
Can we use `ftell` and fail if the result is not `0`? If not, can we use `rewind` instead?

w2yehia replied:
will use ftell.

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,113 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+
+  // Read the entire file into memory.
+  size_t BytesRead = fread(Buffer, 1, Length, F);
+  if (BytesRead != (size_t)Length || ferror(F)) {
+    PROF_ERR("%s: fread failed: %s\n", __func__, strerror(errno));
```
Encountering EOF does not set `errno` AFAIK.

hubert-reinterpretcast replied:
POSIX (but not C) `fread` sets `errno` for the error case (EOF not being an error case). The check for `BytesRead` is sufficient for checking for EOF and for checking for errors.

w2yehia replied:
this is the latest version:
```
if (BytesRead != (size_t)Length) {
    PROF_ERR("%s: fread failed%s\n", __func__,
             feof(F) ? ": end of file reached" : "");
    free(Buffer);
    return;
}
```
let me know if you have any comments or this conversation can be resolved. Thanks.

hubert-reinterpretcast replied:
Could technically do
```
errno = 0;
...
... (errno != 0 ? strerror(errno) : "") ...
```
but that seems overkill. Feel free to mark the conversation resolved.

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,113 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+  size_t BytesRead = fread(Buffer, 1, Length, F);
+  if (BytesRead != (size_t)Length || ferror(F)) {
+    PROF_ERR("%s: fread failed: %s\n", __func__, strerror(errno));
+    return;
```
Might be good to `free(Buffer)` or have a comment in the code why that's not needed.

w2yehia replied:
same as above.

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,113 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+  }
+  if (fseek(F, 0L, SEEK_SET) != 0) {
+    PROF_ERR("%s: fseek(0, SEEK_SET) failed: %s\n", __func__, strerror(errno));
+    return;
```
Might be good to `free(Buffer)` or have a comment in the code why that's not needed.

w2yehia replied:
will free the Buffer, even though we're not recovering from this error in any meaningful way, and the user program is in the midst of a shut down.

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,113 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+    PROF_ERR("%s: Buffer has invalid state: %d", __func__, Buf->Status);
+    break;
+  }
+}
```
Would be good to clear `Buf` here.

w2yehia replied:
will do:
```
  Buf->Addr = NULL;
  Buf->Status = MM_INVALID;
```

w2yehia commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,113 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+                                                      ManagedMemory *Buf) {
+  Buf->Status = MM_INVALID;
+
+  if (!F || isMmapSafe(fileno(F))) {
```
> Can we simply return if !F?
yes.

> Please note that the caller is somewhat expecting this function to set errno to non-zero on any failure (maybe it shouldn't).
good observation. I'll report mmap's error in this function (the other calls are already error checked), and I'll change the caller (`mmapProfileForMerging`) so that it doesn't attempt reading errno.

w2yehia replied:
Unless you insist otherwise, I'll just not check for `!F` and assume it's false because many existing functions that expect `F != 0` don't check for it,

hubert-reinterpretcast replied:
> Unless you insist otherwise, I'll just not check for `!F` and assume it's false because many existing functions that expect `F != 0` don't check for it,

I think the old code has undefined behaviour for the `!F` case and the new code is not worse in that aspect.

Commit:
code review

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,118 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+      _Static_assert(sizeof(Vfs.f_fsid) == sizeof(Vp->vmt_fsid),
+                     "fsid length mismatch");
+      if (memcmp(&Vfs.f_fsid, &Vp->vmt_fsid, sizeof Vfs.f_fsid) == 0) {
+        free(Buf);
+        return (Vp->vmt_flags & MNT_REMOTE) == 0;
```
Looks like the buffer is being read from after it's freed?
```suggestion
        int Answer = (Vp->vmt_flags & MNT_REMOTE) == 0;
        free(Buf);
        return Answer;
```

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,118 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+  size_t BytesRead = fread(Buffer, 1, Length, F);
+  if (BytesRead != (size_t)Length) {
+    PROF_ERR("%s: fread failed%s\n", __func__,
+             feof(F) ? ", end of file reached" : "");
```
Minor nit: should be a colon or a semi-colon (not a comma).

w2yehia committed:
Update compiler-rt/lib/profile/InstrProfilingUtil.c

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,118 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+    munmap(Buf->Addr, Length);
+    break;
+  case MM_INVALID:
+    PROF_ERR("%s: Buffer has invalid state: %d", __func__, Buf->Status);
```
```suggestion
    PROF_ERR("%s: Buffer has invalid state: %d\n", __func__, Buf->Status);
```

w2yehia commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,119 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+  size_t BytesRead = fread(Buffer, 1, Length, F);
+  if (BytesRead != (size_t)Length) {
+    PROF_ERR("%s: fread failed%s\n", __func__,
+             feof(F) ? ", end of file reached" : "");
```
```suggestion
             feof(F) ? ": end of file reached" : "");
```

w2yehia committed:
Update compiler-rt/lib/profile/InstrProfilingUtil.c

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,118 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+  }
+
+  if (getenv("LLVM_PROFILE_VERBOSE"))
+    PROF_NOTE("Could not use mmap; using fread instead.%s\n", "");
```
Minor nit: The other case I saw used this pattern instead of using an empty string for `%s`:
```suggestion
    PROF_NOTE("%s\n", "Could not use mmap; using fread instead.");
```

w2yehia committed:
Update compiler-rt/lib/profile/InstrProfilingUtil.c

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>

w2yehia committed:
Update compiler-rt/lib/profile/InstrProfilingUtil.c

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,119 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+  (void)&is_local_filesystem; // a fake reference to satisfy -Wunused-function
+#ifdef _AIX
+  return is_local_filesystem(Fd);
+#endif
+  return 1;
```
```suggestion
#else
  return 1;
#endif
```

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,119 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+static int isMmapSafe(int Fd) {
+  if (getenv("LLVM_NO_MMAP")) // For testing purposes.
+    return 0;
+  (void)&is_local_filesystem; // a fake reference to satisfy -Wunused-function
```
Consider using `__attribute__((unused))` (preferred spelling: `__attribute__((__unused__))`) via an `UNUSED` macro (like for other compiler-rt libraries) on the function definition instead.

```suggestion
```

w2yehia replied:
Thanks. Done

w2yehia committed:
Update compiler-rt/lib/profile/InstrProfilingUtil.c

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,120 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+  int Tries = 3;
+  while (Tries--) {
+    Buf = malloc(BufSize);
+    Ret = mntctl(MCTL_QUERY, BufSize, Buf);
```
```suggestion
    // mntctl returns -1 if `Buf` is `NULL`.
    Ret = mntctl(MCTL_QUERY, BufSize, Buf);
```

hubert-reinterpretcast commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,120 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+  case MM_MMAP:
+    munmap(Buf->Addr, Length);
+    break;
+  case MM_INVALID:
```
I don't feel strongly about this, but:
```suggestion
  default:
```

hubert-reinterpretcast commented on compiler-rt/test/profile/instrprof-no-mmap-during-merging.c:
```diff
@@ -0,0 +1,25 @@
+// RUN: rm -f *.profraw
+// RUN: %clang_pgogen %s -o a.out
+
+// Need to run a.out twice, the second time a merge will occur which will trigger an mmap.
```
```suggestion
// Need to run a.out twice. On the second time, a merge will occur, which will
// trigger an mmap.
```

hubert-reinterpretcast commented on compiler-rt/test/profile/Posix/instrprof-fork.c:
```diff
@@ -16,17 +20,21 @@ __attribute__((noinline)) void func1() {}
+    }                   //             |
+  }                     // ------------+------------
+  int status;           //   20     10 |   1     0
+  wait(&status);        // (*)  the child inherits counter values prior to fork
```
```suggestion
  i = 10;
  while (i-- > 0)
    wait(&status);      // (*)  the child inherits counter values prior to fork
```

w2yehia committed:
Update compiler-rt/lib/profile/InstrProfilingUtil.c

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>

w2yehia committed:
Update compiler-rt/lib/profile/InstrProfilingUtil.c

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>

w2yehia committed:
Update compiler-rt/lib/profile/InstrProfilingUtil.c

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>

w2yehia committed:
Update compiler-rt/test/profile/instrprof-no-mmap-during-merging.c

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>

w2yehia committed:
Update compiler-rt/test/profile/Posix/instrprof-fork.c

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>

Commit:
define COMPILER_RT_UNUSED

hubert-reinterpretcast commented on compiler-rt/test/profile/Posix/instrprof-fork.c:
```diff
@@ -16,17 +20,23 @@ __attribute__((noinline)) void func1() {}
+    }                   //             |
+  }                     // ------------+------------
+  int status;           //   20     10 |   1     0
+  i = 10;
+  while (i-- > 0)
+    wait(&status);      // (*)  the child inherits counter values prior to fork
+  return 0;             //      from the parent in non-continuous mode.
```
My bad for splitting the comment block in my previous suggestion.
```suggestion
  i = 10;               // (*)  the child inherits counter values prior to fork
  while (i-- > 0)       //      from the parent in non-continuous mode.
    wait(&status);
  return 0;
```

hubert-reinterpretcast reviewed (approved):
LGTM with some suggestions.

petrhosek commented on compiler-rt/lib/profile/InstrProfilingUtil.h:
```diff
@@ -31,6 +31,25 @@ int lprofUnlockFileHandle(FILE *F);
 FILE *lprofOpenFileEx(const char *Filename);
+
+enum MemoryStatus {
+  MM_INVALID = 0x1, // Addr is not a valid address
+  MM_MMAP = 0x2,    // Addr was mmap'ed
+  MM_MALLOC = 0x4   // Addr was malloc'ed
```
What does `MM` stand for? Given the `enum` is called `MemoryStatus`, I'd expect these values to use prefix `MS_` as is the case elsewhere in LLVM.

Also, is there any reason why the values are defined as bit flags? As far as I can tell, we never want these to be combined (i.e. no `MM_MMAP | MM_MALLOC`) so the values can use natural ordering.

petrhosek commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,120 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
 
+// Return 1 (true) if the file descriptor Fd represents a file that is on a
+// local filesystem, otherwise return 0.
+COMPILER_RT_UNUSED static int is_local_filesystem(int Fd) {
```
Can we avoid using a mixture of `snake_case` and `camelCase` for function names (i.e. `is_local_filesystem` vs `isMmapSafe`)?

w2yehia replied:
done

petrhosek commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,120 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+static int isMmapSafe(int Fd) {
+  if (getenv("LLVM_NO_MMAP")) // For testing purposes.
+    return 0;
+#ifdef _AIX
+  return is_local_filesystem(Fd);
+#else
+  return 1;
+#endif
```
Wouldn't using `mmap` be an issue even on other platforms when the profile is stored on NFS (e.g. Linux)?

w2yehia replied:
I couldn't find any evidence that it's broken, in general, on linux or other platforms. 

Note that in this case, we're using mmap to read a file, while maintaining exclusive access to it (opened with `lprofOpenFileEx`) -- we've seen instances, but on AIX only, where we receive a SIGBUS when accessing mmap'ed memory (in clang, while reading a source file) when the file underneath is changed by another process (happens in a makefile build).

No one reported a problem in the past 5+ years, so I think we're good not doing anything on non-AIX platforms for now.

petrhosek commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,120 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+}
+
+static int isMmapSafe(int Fd) {
+  if (getenv("LLVM_NO_MMAP")) // For testing purposes.
```
The environment variable should use `LLVM_PROFILE_` prefix for consistency.

w2yehia committed:
Update compiler-rt/lib/profile/InstrProfilingUtil.c

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>

w2yehia committed:
Update compiler-rt/test/profile/Posix/instrprof-fork.c

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>

david-xl commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,120 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
 
+// Return 1 (true) if the file descriptor Fd represents a file that is on a
+// local filesystem, otherwise return 0.
+COMPILER_RT_UNUSED static int is_local_filesystem(int Fd) {
```
guard the whole function body with #ifdef AIX

w2yehia replied:
done

david-xl commented on compiler-rt/lib/profile/InstrProfilingUtil.c:
```diff
@@ -258,6 +267,120 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
+  return 0;
+}
+
+static int isMmapSafe(int Fd) {
```
Seems better to flip it to 'IsMmapUnsafe' to capture the exceptional case on AIx.

w2yehia replied:
I slightly prefer isMmapSafe to avoid the double negation at the call site (becomes `!isMmapUnsafe`) and the return expression (becomes return 0) in the non-AIX path. 
Unless we rename to isMmapDangerous or isMmapHazardous (sorry cudn't come up with better descriptions).
Let me know if you prefer a change here.

david-xl replied:
sounds good.

w2yehia committed:
Update compiler-rt/lib/profile/InstrProfilingUtil.c

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>

Commit:
code review: rename enumerator names MM_* to MS_*

Commit:
code review: use default enum values instread of bit flags.

With this change, an initialized ManagedMemory object will get the MS_INVALID(==0) status.

Commit:
code review: rename function is_local_filesystem and env var LLVM_NO_MMAP to the correct the conventions

Commit:
fix testcase

w2yehia commented:
@petrhosek @david-xl Thank you for your review.
I've left two unresolved comments of yours for you to close or reply to.

ahatanak commented:
@w2yehia: the tests added in this PR are failing: https://green.lab.llvm.org/job/llvm.org/job/clang-stage1-RA/3899/testReport/junit/Profile-x86_64/Profile-x86_64/instrprof_no_mmap_during_merging_c/

w2yehia commented:
@ahatanak @aeubanks Thanks for notifying and reverting.
I don't have a mac but got a colleague to try out a fix for me. I also couldn't find a clang on macos on godbolt.org.
The failure is due to missing value profiling data.
The `memcpy` calls are not lowered by the FE to the `llvm.memcpy` intrinsic. They are macro expanded to `memcpy_chk` which is not recognized by the value profiling instrumentation.
So memory intrinsics (at least `memcpy`) value profiling is currently broken on macos.
